src/main/java/htsjdk/samtools/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,677 @@
+     * @throws IllegalStateException    if no base was added to the previous sequence or the writer is already closed of
+     *                                  the sequence has been already added.
+     * @throws IOException              if such exception is thrown when writing into the output resources.
+     */
+    public FastaReferenceWriter startSequence(final String sequenceName, final String description, final int basesPerLine)
+            throws IOException {
+        assertIsNotClosed();
+        checkSequenceName(sequenceName);
+        final String nonNullDescription = checkDescription(description);
+        FastaReferenceWriterBuilder.checkBasesPerLine(basesPerLine);
+        closeSequence();
+        if (sequenceNames.contains(sequenceName)) {
+            throw new IllegalStateException("the input sequence name '" + sequenceName + "' has already been added");
+        }
+        currentSequenceName = sequenceName;
+        currentBasesPerLine = basesPerLine;
+        final StringBuilder builder = new StringBuilder(sequenceName.length() + nonNullDescription.length() + 2);
+        builder.append(HEADER_START_CHAR).append(sequenceName);
+        if (!nonNullDescription.isEmpty()) {
+            builder.append(HEADER_NAME_AND_DESCRIPTION_SEPARATOR).append(nonNullDescription);
+        }
+        fastaStream.write(builder.toString().getBytes(CHARSET));
+        fastaStream.write(LINE_SEPARATOR);
+        currentSequenceOffset = fastaStream.getBytesWritten();
+        return this;
+    }
+
+    private void closeSequence()
+            throws IOException {
+        if (currentSequenceName != null) {
+            if (currentBasesCount == 0) {
+                throw new IllegalStateException("no base was added");
+            }
+            sequenceNames.add(currentSequenceName);
+            writeIndexEntry();
+            writeDictEntry();
+            fastaStream.write(LINE_SEPARATOR);
+            currentBasesCount = 0;
+            currentLineBasesCount = 0;
+            currentSequenceName = null;
+        }
+    }
+
+    private void writeIndexEntry()
+            throws IOException {
+        indexWriter.append(currentSequenceName).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesCount)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentSequenceOffset)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesPerLine)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesPerLine + LINE_SEPARATOR.length)).append(LINE_SEPARATOR_CHR);
+    }
+
+    private void writeDictEntry() {
+        dictCodec.encodeSequenceRecord(new SAMSequenceRecord(currentSequenceName, (int) currentBasesCount));
+    }
+
+    /**
+     * Adds bases to current sequence from a {@code byte} array.
+     *
+     * @param basesString String containing the bases to be added.
```

Comment:
add: string will be interpreted using ascii and will throw if any character is > 128.